### PR TITLE
Introduce a public API to publish custom components

### PR DESCRIPTION
**Pull request by melix**
### Context

This PR introduces a public API to publish _simple_ custom components. As a proof-of-concept, it migrates the Java plugin to use this, and if the PR is accepted, migrating the Java Platform plugin to it should be trivial.

The idea is that lots of publications do not need a rich component model. Instead, we can rely on the existing _outgoing configurations_ to produce components. This is what the Java plugin does, and I have verified that other plugins like the _Shadow_ plugin would also benefit from this.

Therefore, it introduces a `SoftwareComponentFactory`, which can be injected in a plugin like this:

```
class MyPlugin implements Plugin<Project> 
   @Inject
    MyPlugin(SoftwareComponentFactory factory) { ... }
```

then the plugin can create adhoc software components, to which outgoing configurations can be attached. The API of an adhoc component is very simple:

```
void addVariantsFromConfiguration(
        Configuration outgoingConfiguration, 
        Action<? super ConfigurationVariantDetails> action);
```

then using it in a plugin is done like this:

```
AdhocComponentWithVariants java = softwareComponentFactory.adhoc("java");
        java.addVariantsFromConfiguration(configurations.getByName(API_ELEMENTS_CONFIGURATION_NAME), new JavaConfigurationVariantMapping("compile", false));
        java.addVariantsFromConfiguration(configurations.getByName(RUNTIME_ELEMENTS_CONFIGURATION_NAME), new JavaConfigurationVariantMapping("runtime", false));
        project.getComponents().add(java);
```

The `details` object passed to the `addVariantsFromConfiguration` serves 2 purposes:

1. filtering which outgoing `ConfigurationVariant` are going to be published. For example, we don't want to publish the variants which are a "classes directory" (since it's technically not doable)/
2. determining how those variants can be mapped to Maven POMs (select a scope and if the dependencies should be published as optional)

The Maven mapping here is a tradeoff, to avoid coupling `SoftwareComponent` to Maven.

### CODE REVIEW

**Comment by big-guy** [comment]
@lacasseio FYI, this is something we should consider for the native plugins

**Comment by melix** [comment]
@big-guy 
> @lacasseio FYI, this is something we should consider for the native plugins

I took a lot at what the Kotlin native does too. It's more complex in your case since the component model is driving everything. So you have to create a component, which can be configured and _then_ will generate tasks, ... In my case this is much simpler since the _plugin_ is creating tasks, configurations, and eventually the component model is derived _just_ for publication.

I think one way to workaround would be to have a "secondary component" that is only used for publishing, so the "main component" would be driving the model and you'd have another, adhoc component for publishing. I _think_ this one could use exactly the same API as here, but I'm not familiar enough with the native plugin to tell.

**Comment by technoir42** [comment]
This is great! Hopefully Android Gradle Plugin will leverage this to let us publish AAR libraries with `maven-publish` out of the box.
So far I've been using https://github.com/wupdigital/android-maven-publish but it relies on internal APIs and as the result it breaks with almost every Gradle release.

**Comment by melix** [comment]
Indeed from what I read [here](https://github.com/wupdigital/android-maven-publish/blob/master/plugin/src/main/groovy/digital/wup/android_maven_publish/AndroidVariantLibrary.groovy) it would be fairly straightforward to use this API instead.

File: subprojects/plugins/src/main/java/org/gradle/api/internal/java/JavaLibrary.java [hunk 1]
```diff
@@ -112,87 +106,21 @@ private UsageContext createCompileUsageContext() {
     }
 
     @Override
-    public void addFeatureVariantsFromConfiguration(Configuration outgoingConfiguration, Spec<? super ConfigurationVariant> spec) {
+    public void addFeatureVariantsFromConfiguration(Configuration outgoingConfiguration, final Spec<? super ConfigurationVariant> spec) {
```
Comment (ljacomet) [inline subprojects/plugins/src/main/java/org/gradle/api/internal/java/JavaLibrary.java]: ⭕️ Given the changes in this PR, the recently introduced `ComponentWithFeatures` can be removed completely.

Comment (melix) [reply subprojects/plugins/src/main/java/org/gradle/api/internal/java/JavaLibrary.java]: Removed `ComponentWithFeatures`, good catch!

**Pull request merged**